### PR TITLE
Fix class weighting for missing classes.

### PR DESCRIPTION
**Pull request by VahidooX**

### CODE REVIEW

File: examples/nlp/token_classification/punctuation_capitalization.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def create_pipeline(
         if args.use_weighted_loss_punct:
             logging.info(f"Using weighted loss for punctuation task")
             punct_label_freqs = data_layer.dataset.punct_label_frequencies
-            class_weights = calc_class_weights(punct_label_freqs)
+            class_weights = calc_class_weights(punct_label_freqs, max_id=-1)
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuation_capitalization.py]: max_id=-1 is the default, please revert the changes in the script

File: examples/nlp/intent_detection_slot_tagging/joint_intent_slot_with_bert.py [hunk 2]
```diff
@@ -99,6 +99,7 @@
     create_tb_writer=True,
     files_to_copy=[__file__],
     add_time_to_log_dir=True,
+    placement=nemo.core.DeviceType.CPU,
```
Comment (ekmb) [inline examples/nlp/intent_detection_slot_tagging/joint_intent_slot_with_bert.py]: why `placement=nemo.core.DeviceType.CPU` ?

File: examples/nlp/token_classification/token_classification.py [hunk 3]
```diff
@@ -194,11 +195,7 @@ def create_pipeline(
         if args.use_weighted_loss:
             logging.info(f"Using weighted loss")
             label_freqs = data_layer.dataset.label_frequencies
-            class_weights = nemo.collections.nlp.data.datasets.datasets_utils.data_preprocessing.calc_class_weights(
-                label_freqs
-            )
-
-            logging.info(f"class_weights: {class_weights}")
+            class_weights = calc_class_weights(label_freqs, max_id=-1)
```
Comment (ekmb) [inline examples/nlp/token_classification/token_classification.py]: max_id=-1 is the default, please revert the changes in the script

File: nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py [hunk 4]
```diff
@@ -77,7 +79,11 @@ def get_label_stats(labels, outfile='stats.tsv'):
         if i < 3:
             logging.info(f'{i} item: {k}, {v} out of {total}, {v / total}.')
         i += 1
-    return total, label_frequencies
+
+    freq_dict = {}
+    for lf in label_frequencies:
+        freq_dict[lf[0]] = lf[1]
```
Comment (ekmb) [inline nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py]: you can add this to the previous for loop

File: nemo/collections/nlp/data/datasets/joint_intent_slot_dataset/joint_intent_slot_descriptor.py [hunk 5]
```diff
@@ -72,6 +79,9 @@ def __init__(self, data_dir, none_slot_label='O', pad_label=-1):
         self.slots_label_ids = JointIntentSlotDataDesc.label2idx(self.slot_dict_file)
         self.num_slots = len(self.slots_label_ids)
 
+        infold = self.data_dir
+        max_slot_id = 0
+        max_intent_id = 0
```
Comment (ekmb) [inline nemo/collections/nlp/data/datasets/joint_intent_slot_dataset/joint_intent_slot_descriptor.py]: you have access to the slot/intent dicts, are you sure you need this?

File: examples/nlp/token_classification/punctuation_capitalization.py (hunk 1 above)
Comment (VahidooX) [reply examples/nlp/token_classification/punctuation_capitalization.py]: When max_id=-1 is passed. It can not handle missing labels. To handle misiing labels, maximum label id should get passed. I put it there to remind us that it does not handle those label missing cases.

File: examples/nlp/intent_detection_slot_tagging/joint_intent_slot_with_bert.py (hunk 2 above)
Comment (VahidooX) [reply examples/nlp/intent_detection_slot_tagging/joint_intent_slot_with_bert.py]: fixed that. I was testing on my laptop. It does not run on cpu without that.

File: nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py (hunk 4 above)
Comment (VahidooX) [reply nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py]: Merged them.

File: examples/nlp/token_classification/punctuation_capitalization.py (hunk 1 above)
Comment (ekmb) [reply examples/nlp/token_classification/punctuation_capitalization.py]: this script doesn't use descriptor and the dict with labels and labels ids is generated from the training data and passed to the eval layer, I don't think that this max_id is relevant here.

Comment (VahidooX) [reply examples/nlp/token_classification/punctuation_capitalization.py]: Removed it.

File: examples/nlp/token_classification/token_classification.py (hunk 3 above)
Comment (VahidooX) [reply examples/nlp/token_classification/token_classification.py]: removed.

File: nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py [hunk 6]
```diff
@@ -314,19 +319,32 @@ def dataset_to_ids(dataset, tokenizer, cache_ids=False, add_bos_eos=True):
...
-    Here label_freq is assumed to be sorted by the frequency. I.e.
-    label_freq[0] is the most frequent element.
+    total_size = 0
+    for lf in label_freq.values():
+        total_size += lf
+    weighted_slots = {label: (total_size / freq) for label, freq in label_freq.items()}
+    return weighted_slots
+
+
+def fill_class_weights(weights, max_id=-1):
```
Comment (ekmb) [inline nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py]: could you please add docstings to this one?

File: nemo/collections/nlp/data/datasets/joint_intent_slot_dataset/joint_intent_slot_descriptor.py (hunk 5 above)
Comment (VahidooX) [reply nemo/collections/nlp/data/datasets/joint_intent_slot_dataset/joint_intent_slot_descriptor.py]: Used num_slots and num_intents instead.

File: nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py (hunk 6 above)
Comment (VahidooX) [reply nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py]: Added. Also set weights for missing classes to one. It was zero before that. What do you think?

Comment (ekmb) [reply nemo/collections/nlp/data/datasets/datasets_utils/data_preprocessing.py]: I think it shouldn't matter since these labels are missing in the training data
